runtime/grid_reduction.cu

```diff
@@ -667,4 +667,243 @@ __device__ void serialReductionStep(
+  // every block, and every thread marked as true. Threads in dimensions marked
+  // as false can be used to parallelize the reduction.
+
+  // Find the reduction id of the participating threads
+  const auto block_reduction_segment_idx =
+      index_utils::maskedOffset<X_THREAD, Y_THREAD, Z_THREAD>(
+          threadIdx, blockDim);
+
+  // Find an id associated within a reduction segment for all
+  // "non-participating" threads, which will parallelize the reductions for the
+  // "participating" threads
+  const auto id_in_block_segment =
+      index_utils::maskedOffset<!X_THREAD, !Y_THREAD, !Z_THREAD>(
+          threadIdx, blockDim);
+
+  // index into iteration dim
+  const auto thread_offset =
+      index_utils::maskedOffset<X_THREAD, Y_THREAD, Z_THREAD>(
+          threadIdx, blockDim);
+
+  // Stride by the "non-participating" threads
+  const auto input_stride_for_thread_in_segment =
+      index_utils::maskedSize<!X_THREAD, !Y_THREAD, !Z_THREAD>(blockDim);
+
+  // T inp = init_val;
+  T inp[vec_size];
+#pragma unroll
+  for (int i = 0; i < vec_size; i++) {
+    inp[i] = init_val;
+  }
+  // Block stride across the reduction until we only have one value per thread
+  for (nvfuser_index_t reduction_i = id_in_block_segment;
+       reduction_i < grid_reduction_segment_size;
+       reduction_i += input_stride_for_thread_in_segment) {
+    // index to global memory
+    auto work_buf_offset =
+        reduction_i * (block_reduction_segment_size * grid_segment_size) +
+        block_reduction_segment_size * idx_in_grid_segment + thread_offset;
+    work_buf_offset *= vec_size;
+
+    // vectorized load from global memory
+    constexpr int total_bytes = vec_size * sizeof(T);
+    constexpr int n_loads = total_bytes <= 16 ? 1 : total_bytes / 16;
+    constexpr int n_elements = total_bytes <= 16 ? vec_size : 16 / sizeof(T);
+#pragma unroll
+    for (int i = 0; i < n_loads; i++) {
+      int i_offset = i * n_elements;
+      T in_reg[n_elements];
+      loadGlobalToLocal<T, n_elements, true, CacheOp::Global>(
+          &in_reg[0], const_cast<T*>(in + work_buf_offset + i_offset));
+#pragma unroll
+      for (int j = 0; j < n_elements; j++) {
+        reduction_op(inp[i_offset + j], in_reg[j]);
+      }
+    }
+  }
+
+  // Block reduce the per thread values into per "participating" thread values
+  // T inp_tmp = init_val;
+  T inp_tmp[vec_size];
+#pragma unroll
+  for (int i = 0; i < vec_size; i++) {
+    inp_tmp[i] = init_val;
+  }
+  blockIterGroupedReduce<!X_THREAD, !Y_THREAD, !Z_THREAD, Aligned, vec_size>(
+      inp_tmp, inp, reduction_op, shared_buf, true, init_val);
+  const bool should_write = (X_THREAD || threadIdx.x == 0) &&
+      (Y_THREAD || threadIdx.y == 0) && (Z_THREAD || threadIdx.z == 0);
+  if (should_write && write_pred) {
+#pragma unroll
+    for (int i = 0; i < vec_size; i++) {
+      reduction_op(out[i], inp_tmp[i]);
+    }
+  }
+}
+
+template <
```

Comment:
nitpick: understandably we can refer to `gridReduce` for the main part of documentation. We should put a quick note here on how this method differs from that.

Reply:
Especially comment on what the added argument does. i.e. `vec_size`.

Reply:
added